task/host-sp-comms/src/main.rs

```diff
@@ -701,13 +710,24 @@ impl ServerImpl {
 
         // We defer any actions until after we've serialized our response to
         // avoid borrow checker issues with calling methods on `self`.
+        //
+        // NOTE: setting any action will also clear the `host_boot_just_failed`
+        // flag, to avoid keeping incorrect state around. Since we have no
+        // explicit way of tracking the host boot software's lifecycle, this is
+        // a compromise.
         let mut action = None;
+
         let response = match request {
             HostToSp::_Unused => {
                 Some(SpToHost::DecodeFailure(DecodeFailureReason::Deserialize))
             }
             HostToSp::RequestReboot => {
-                action = Some(Action::RebootHost);
+                if core::mem::take(&mut self.host_boot_just_failed) {
+                    // Do _not_ reboot. Reinterpret this as a poweroff request.
+                    action = Some(Action::PowerOffHost);
```

Comment:
I'm happy to look at this more tomorrow, but we might need to address the `TODO` I left in `handle_jefe_notification`. There may be a sequence of reboot operations where we still reboot even if we set this action.

Reply:
Good to know, let's talk tomorrow.

Reply:
I think I have convinced myself this is okay (where "this" is really "the handling of `self.reboot_state`"):

1. `self.reboot_state` starts out `None`
2. It can only go from `None` to `Some(_)` inside a call to `power_off_host(true)`
3. Once it goes to `Some(_)`, it will eventually cause the host to come back on through one of these paths:
  - If `power_off_host(true)` succeeds in asking the sequencer to go to A2, it's set to `Some(WaitingForA2)`. Once we get a notification from jefe that we've entered A2, it will change to `Some(WaitingInA2RebootDelay)` and start the `Timers::WaitingInA2ToReboot` timer. Once that timer fires, we'll enter `handle_reboot_waiting_in_a2_timer`, set the state to A0, and set `self.reboot_state` back to `None.
  - If `power_off_host(true)` fails to ask the sequencer to go to A2 because we're already in A2, it's set to `Some(WaitingInA2RebootDelay)` and we start the `Timers::WaitingInA2ToReboot` timer. Once that timer fires, we'll enter `handle_reboot_waiting_in_a2_timer`, set the state to A0, and set `self.reboot_state` back to `None.

`power_off_host(true)` is only called in two places: if we get a jefe notification that we've gone to `A0Reset`, or if the host sends us `HostToSp::RequestReboot`. If either of those happened and before we go through the above timer-based process (~5 seconds) to reboot the host we landed here with a host boot fail, the `PowerOffHost` would be effectively ignored because we'd already be in the process of rebooting the host. But that doesn't seem likely, and if it happens it's maybe fine to honor the "first request" that triggered the reboot?

Reply:
Currently, if the host sends back-to-back "power off" and "reboot" requests, we'll behave ... rather strangely. I wonder if it might be worth adding code to stop listening to IPCC entirely starting at any reboot/powerdown request and continuing up until we believe the host to have been powered back on.